### ISSUE TITLE
Performance vs Clarity

### ISSUE THREAD

**Issue by davidmascharka** [davidmascharka/MyNN#13]
We should discuss the mertis of a performance/clarity tradeoff in the library. We can take advantage of some of the math functions in MyGrad to write some neural network utilities incredibly clearly and concisely, at the cost of some performance hit.

For example, [L1 before](https://github.com/davidmascharka/MyNN/blob/6b9e045998776e514edb931c8ff1092ca803eac7/mynn/losses/l1_loss.py) and [L1 after](https://github.com/davidmascharka/MyNN/blob/master/mynn/losses/l1_loss.py). It's immediately obvious just looking at the current L1 implementation what it's doing; all you need is to understand how the primitives backprop. If you know that (or trust that they are) then you don't need to see details. 

Another benefit of using the MyGrad primitives where applicable is the fact that we stay in step if MyGrad ops get rewritten; we don't need to update both implementations if anything changes for some reason.

However, since you're relying on the MyGrad primitives to perform the backward pass on its own, you lose out on some performance benefits of writing these operations specifically. Some of the operations incur about a 2x slowdown versus writing the forward and backward explicitly, which is not ideal. For example, I believe [log softmax](https://github.com/davidmascharka/MyNN/blob/master/mynn/activations/log_softmax.py) was about twice as slow, I believe.

Currently, my strategy is to write all operations using MyGrad primitives that we can. This helps improve development speed. My initial idea is that once we get more fully-featured we can perfom some analysis of where slowdowns are coming from and focus on optimizing those. There's little point in optimizing an op if it's not used very much or if its overhead versus another op is miniscule.

Thoughts on this tradeoff?
